imp_developer.py

```diff
@@ -1305,55 +2044,217 @@ def __init__(self, env):
         self.env = env
         self.poll_url = None
         self.last_shown_log = None
+        self.sock = None
+        self.devices = []
+        self.has_logs = False
+        # Supported states:
+        self.IDLE = 0
+        self.INIT = 1
+        self.POLL = 2
+        self.FAIL = 3
+        # set initial state
+        self.state = self.IDLE
+        # prevent multiple requests when http is pending
+        self.update_log_started = False
+        # no timer on start-up
+        self.keep_alive = None
+
+    def start(self):
+        if self.state == self.IDLE:
+            self.state = self.INIT
+            self.write_to_console("Log stream start requested ...\n")
+        else:
+            log_debug("Unexpected log manger state")
+
+    def stop(self, is_restart=False):
+        if self.state == self.POLL:
+            if is_restart:
+                self.write_to_console("Real-time logging restart requested.")
+                # set current state to idle
+                self.state = self.IDLE
+                # and then trigger log start
+                sublime.set_timeout_async(lambda: update_log_windows(False), 0)
+                return
+            # stop logs
+            self.write_to_console("Real-time logging has stopped. Please refresh to enable it again.")
+        elif self.state == self.INIT:
+            self.write_to_console("Real-time logging not started. Please refresh to enable it again.")
+
+        # change current state depends on initial
+        self.state = self.IDLE
+    #
+    def check_imp_error(self, error):
+        if not error:
+            return False
+
+        # Handle invalid credentials use-case only
+        # the following string should restart logs via command
+        # which lead to access token renew
+        if error["code"] == ImpRequest.INVALID_CREDENTIALS:
+            sublime.set_timeout_async(
+                lambda: self.env.window.run_command("imp_show_console", {"cmd_on_complete": "auth"}), 0)
+
+        return True
+
+
+    def __read_logs(self):
+        if self.sock and type(self.sock) != None and self.sock.fp != None:
+            next_log = False
+            next_cmd = False
+            logs = []
+            count = PL_LOGS_MAX_PER_REQUEST
+            while count > 0:
+                # lets read no more than coun logs per one loop
+                rd, wd, ed = select.select([self.sock], [], [], 0)
+                if not rd:
+                    break
+                else:
+                    for line in self.sock:
+                        if next_log:
+                            next_log = False
+                            logs.append(line.decode("utf-8"))
+                        elif next_cmd:
+                            next_cmd = False
+                            if line == b'data: closed\n':
+                                self.reset()
```

Comment:
`self.keep_alive = None` is missing here!

Reply:
it happens inside reset method